File: backend_InMoment/InMoment.Domain/Notifications/NotificationSettings.cs
using InMoment.Domain.Common;

namespace InMoment.Domain.Notifications;

public sealed class NotificationSettings : Entity<Guid>
{
    public Guid UserId { get; private set; }

    public bool PushEnabled { get; private set; }

    public bool PushGroupInvitations { get; private set; }
    public bool PushReactions { get; private set; }
    public bool PushComments { get; private set; }
    public bool PushReplies { get; private set; }
    public bool PushMentions { get; private set; }
    public bool PushPosts { get; private set; }

    public bool PushRetention { get; private set; }
    public bool PushProductUpdates { get; private set; }

    public DateTime CreatedAtUtc { get; private set; }
    public DateTime UpdatedAtUtc { get; private set; }

    private NotificationSettings() { }

    public static NotificationSettings CreateDefault(Guid userId)
    {
        if (userId == Guid.Empty)
            throw new ValidationException("UserId is required.");

        var now = DateTime.UtcNow;

        return new NotificationSettings
        {
            Id = Guid.NewGuid(),
            UserId = userId,
            PushEnabled = true,
            PushGroupInvitations = true,
            PushReactions = true,
            PushComments = true,
            PushReplies = true,
            PushMentions = true,
            PushPosts = true,
            PushRetention = true,
            PushProductUpdates = true,
            CreatedAtUtc = now,
            UpdatedAtUtc = now
        };
    }

    public void Update(
        bool pushEnabled,
        bool pushGroupInvitations,
        bool pushReactions,
        bool pushComments,
        bool pushReplies,
        bool pushMentions,
        bool pushPosts,
        bool pushRetention,
        bool pushProductUpdates)
    {
        PushEnabled = pushEnabled;
        PushGroupInvitations = pushGroupInvitations;
        PushReactions = pushReactions;
        PushComments = pushComments;
        PushReplies = pushReplies;
        PushMentions = pushMentions;
        PushPosts = pushPosts;
        PushRetention = pushRetention;
        PushProductUpdates = pushProductUpdates;
        UpdatedAtUtc = DateTime.UtcNow;
    }

    public bool IsPushEnabledFor(NotificationType type)
    {
        if (!PushEnabled)
            return false;

        return type switch
        {
            NotificationType.GroupInvitationReceived => PushGroupInvitations,
            NotificationType.ReactionOnPhoto => PushReactions,
            NotificationType.CommentOnPhoto => PushComments,
            NotificationType.ReplyToComment => PushReplies,
            NotificationType.CommentMention => PushMentions,
            NotificationType.PhotoPublishedInGroup => PushPosts,

            NotificationType.ShareReminder => PushRetention,
            NotificationType.FeedbackPrompt => PushRetention,
            NotificationType.Anniversary => PushRetention,
            NotificationType.SystemMemoryReady => PushRetention,
            NotificationType.ModeratorAnnouncement => PushProductUpdates,
            NotificationType.ProductAnnouncement => PushProductUpdates,

            NotificationType.ReportReviewed => true,
            NotificationType.ReportAppealSubmitted => true,
            NotificationType.ReportAppealReviewed => true,

            _ => false
        };
    }
}

File: frontend_InMoment/lib/features/notifications/models/notification_settings.dart
class NotificationSettingsModel {
  final bool pushEnabled;
  final bool pushGroupInvitations;
  final bool pushReactions;
  final bool pushComments;
  final bool pushReplies;
  final bool pushMentions;
  final bool pushPosts;
  final bool pushRetention;
  final bool pushProductUpdates;
  final DateTime? createdAtUtc;
  final DateTime? updatedAtUtc;

  const NotificationSettingsModel({
    required this.pushEnabled,
    required this.pushGroupInvitations,
    required this.pushReactions,
    required this.pushComments,
    required this.pushReplies,
    required this.pushMentions,
    required this.pushPosts,
    required this.pushRetention,
    required this.pushProductUpdates,
    this.createdAtUtc,
    this.updatedAtUtc,
  });

  factory NotificationSettingsModel.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic value) {
      if (value == null) return null;
      return DateTime.tryParse(value.toString());
    }

    return NotificationSettingsModel(
      pushEnabled: json['pushEnabled'] as bool? ?? true,
      pushGroupInvitations: json['pushGroupInvitations'] as bool? ?? true,
      pushReactions: json['pushReactions'] as bool? ?? true,
      pushComments: json['pushComments'] as bool? ?? true,
      pushReplies: json['pushReplies'] as bool? ?? true,
      pushMentions: json['pushMentions'] as bool? ?? true,
      pushPosts: json['pushPosts'] as bool? ?? true,
      pushRetention: json['pushRetention'] as bool? ?? true,
      pushProductUpdates: json['pushProductUpdates'] as bool? ?? true,
      createdAtUtc: parseDate(json['createdAtUtc']),
      updatedAtUtc: parseDate(json['updatedAtUtc']),
    );
  }

  NotificationSettingsModel copyWith({
    bool? pushEnabled,
    bool? pushGroupInvitations,
    bool? pushReactions,
    bool? pushComments,
    bool? pushReplies,
    bool? pushMentions,
    bool? pushPosts,
    bool? pushRetention,
    bool? pushProductUpdates,
    DateTime? createdAtUtc,
    DateTime? updatedAtUtc,
  }) {
    return NotificationSettingsModel(
      pushEnabled: pushEnabled ?? this.pushEnabled,
      pushGroupInvitations:
          pushGroupInvitations ?? this.pushGroupInvitations,
      pushReactions: pushReactions ?? this.pushReactions,
      pushComments: pushComments ?? this.pushComments,
      pushReplies: pushReplies ?? this.pushReplies,
      pushMentions: pushMentions ?? this.pushMentions,
      pushPosts: pushPosts ?? this.pushPosts,
      pushRetention: pushRetention ?? this.pushRetention,
      pushProductUpdates: pushProductUpdates ?? this.pushProductUpdates,
      createdAtUtc: createdAtUtc ?? this.createdAtUtc,
      updatedAtUtc: updatedAtUtc ?? this.updatedAtUtc,
    );
  }
}